### DISCUSSION TITLE
calculate fast.farm output power

### DISCUSSION THREAD

**Discussion by clun1** [OpenFAST/openfast#1553] [General]
I would like to know whether there are power calculation parameters in fast.farm. If there are no relevant parameters, is there any other method to calculate the output power of wind farm

**Comment by jjonkman** [comment]
Dear @clun1,

When running FAST.Farm, you can output the power from each OpenFAST wind turbine model and sum them to get the total output power of the wind farm.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

I want the output power of the wind turbine to be 90% of the original output of the wind turbine, how can I modify diacon.all to control the wind turbine? The following is the wind turbine I control, but the fluctuation is relatively large, hope to get your guidance
![image](https://github.com/OpenFAST/openfast/assets/93849049/a7f5a133-9b4d-4ae2-a62c-02c509cf107a)
![image](https://github.com/OpenFAST/openfast/assets/93849049/a8189894-5909-4cd7-bac8-e57aa7a4b591)

Best regards,

**Reply by jjonkman** [reply]
Dear @clun1,

I'm not sure I fully understand your question.  If you want the output power to change for a given wind turbine, you must change the control strategy, e.g., by changing the target rotor speed and/or blade-pitch angles.

Best regards,

**Comment by clun1** [comment]
Dear jjonkman,

The control strategy I use is to increase the rotor speed of the wind turbine to deviate from the maximum wind energy capture coefficient, as shown in the yellow curve in the picture. However, under turbulent wind conditions, the result of this strategy is that the output power fluctuates greatly, as shown in the figure above. Moreover, this control method is completely different from the control method in DISCON.DLL. If I want to modify the original control mode in discon.dll, so that the output power of the wind turbine is 90% of the original output power of the wind turbine, what should I do, looking forward to your guidance
![image](https://github.com/OpenFAST/openfast/assets/93849049/58f9d3ba-c178-4320-9064-447b632428d6)
![image](https://github.com/OpenFAST/openfast/assets/93849049/f90f7fd7-f6d5-4819-bd1a-5f6cc7eb6eea)

Best regards,

**Reply by jjonkman** [reply]
Dear @clun1,

Are you referring to changing the control in region 2 (maximizing power below rated) or region 3 (fixing power above rated)?

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

Yes, I have tried increasing or decreasing the RPM, but both of these methods cause additional load problems

Best regards,

**Reply by jjonkman** [reply]
Dear @clun1,

I'm sorry, but I'm not sure I understand.  You have not answered my question.  And now you are talking about loads instead of power?  I'm not sure I understand what problem you have and what you want to do.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

Sorry, it's my problem. The control I want to change are region 2 and region 3.

Best regards,

**Reply by dzalkind** [reply]
Hi @clun1,

Here are a few references for controllers that change the power output of the turbine:
- https://www.nrel.gov/docs/fy12osti/54605.pdf
- https://onlinelibrary.wiley.com/doi/full/10.1002/we.2705

In below rated operation, the best way to reduce power is to increase the minimum pitch angle.  Changing the rotor speed set point and resulting TSR does not change the power much but does change the thrust, which it seems like you are experiencing.

In above rated operation, you can reduce the rotor speed set point or the maximum allowed torque.

We have plans to implement something like this in the [ROSCO controller](https://github.com/NREL/ROSCO) in the spring of 2024.

Best, Dan

**Reply by clun1** [reply]
Dear dzalkind,

Thank you very much for your guidance,it help me a lot, thank you!

Best regards,
